File: backend/src/Core/RateLimit.php
<?php
declare(strict_types=1);

namespace Linalysis\Core;

/**
 * Coarse-grained rate limit using a single table. Not as tight as Redis/KV,
 * but good enough for brute-force-protection on auth + extension ingest.
 */
final class RateLimit
{
    /**
     * @throws \RuntimeException if over the limit (caller catches → 429).
     */
    public static function hit(string $key, int $limit, int $windowSeconds): void
    {
        $bucket = floor(time() / $windowSeconds);
        $id = hash('sha256', $key . '|' . $bucket);

        Db::execute(
            "INSERT INTO rate_limits (id, bucket_key, bucket_ts, hits, expires_at)
             VALUES (:id, :k, :ts, 1, DATE_ADD(UTC_TIMESTAMP(), INTERVAL :w SECOND))
             ON DUPLICATE KEY UPDATE hits = hits + 1",
            ['id' => $id, 'k' => $key, 'ts' => (int)$bucket, 'w' => $windowSeconds]
        );

        $count = (int) Db::fetchOne(
            "SELECT hits FROM rate_limits WHERE id = :id",
            ['id' => $id]
        )['hits'];

        if ($count > $limit) {
            throw new RateLimitExceeded(retryAfter: $windowSeconds - (time() % $windowSeconds));
        }
    }
}

final class RateLimitExceeded extends \RuntimeException
{
    public function __construct(public readonly int $retryAfter)
    {
        parent::__construct('Rate limit exceeded.');
    }
}

File: backend-java/src/com/linalysis/auth/RateLimit.java
package com.linalysis.auth;

import com.linalysis.db.Db;
import com.linalysis.util.Password;

import java.sql.SQLException;
import java.util.Map;

/**
 * Coarse bucket counter in MySQL — same table the Cloudflare Worker used.
 * Fine for auth + ingest protection; swap for Redis if traffic grows.
 */
public final class RateLimit {
    private RateLimit() {}

    /** Returns true if request is allowed; false if caller should return 429. */
    public static boolean hit(String key, int limit, int windowSeconds) {
        long bucket = System.currentTimeMillis() / 1000 / windowSeconds;
        String id = Password.sha256Hex(key + "|" + bucket);
        try {
            Db.update(
                "INSERT INTO rate_limits (id, bucket_key, bucket_ts, hits, expires_at) " +
                "VALUES (?, ?, ?, 1, DATE_ADD(UTC_TIMESTAMP(), INTERVAL ? SECOND)) " +
                "ON DUPLICATE KEY UPDATE hits = hits + 1",
                id, key, bucket, windowSeconds);

            Map<String, Object> row = Db.queryOne("SELECT hits FROM rate_limits WHERE id = ?", id);
            if (row == null) return true;
            int hits = ((Number) row.get("hits")).intValue();
            return hits <= limit;
        } catch (SQLException e) {
            // Fail open — we'd rather accept a request than 500 the user
            return true;
        }
    }
}
